src/clients/c++/image_client.cc

```diff
@@ -640,22 +717,43 @@ main(int argc, char** argv)
   // (readdir does not guarantee any particular order).
   std::sort(image_filenames.begin(), image_filenames.end());
 
-  // Preprocess the images into input data according to model
-  // requirements
+  // Create shared memory regions for Input and Output batches
+  if (use_shm) {
+    std::string shm_key = "/input_data";
+    int shm_fd =
+        create_shared_region(shm_key, input_byte_size * image_filenames.size());
+    shm_addr_ip = (float*)(get_shm_addr(
+        shm_fd, 0, input_byte_size * image_filenames.size()));
+    shm_key = "/output_data";
+    shm_fd = create_shared_region(
+        shm_key, output_byte_size * image_filenames.size());
+    shm_addr_op = (float*)(get_shm_addr(
+        shm_fd, 0, output_byte_size * image_filenames.size()));
+  }
+
+  // Preprocess the images into input data according to model requirements and
+  // place in shared memory (if needed)
+  size_t offset = 0;
   std::vector<std::vector<uint8_t>> image_data;
   for (const auto& fn : image_filenames) {
     image_data.emplace_back();
+    size_t byte_size = 0;
     FileToInputData(
-        fn, c, h, w, format, type1, type3, scale, &(image_data.back()));
-
-    if ((image_data.size() == 1) && !preprocess_output_filename.empty()) {
-      std::ofstream output_file(preprocess_output_filename);
-      std::ostream_iterator<uint8_t> output_iterator(output_file);
-      std::copy(image_data[0].begin(), image_data[0].end(), output_iterator);
+        fn, c, h, w, format, type1, type3, scale, &(image_data.back()),
+        shm_addr_ip, offset, &byte_size, use_shm);
+
+    if (!use_shm) {
+      if ((image_data.size() == 1) && !preprocess_output_filename.empty()) {
+        std::ofstream output_file(preprocess_output_filename);
+        std::ostream_iterator<uint8_t> output_iterator(output_file);
+        std::copy(image_data[0].begin(), image_data[0].end(), output_iterator);
+      }
+    } else {
+      offset += byte_size;
     }
   }
 
```

Comment:
Defer shared_memory initialize till here and you don't need many of the above changes. At this point you know the batch size and size of an image (from image_data), so you can allocate the shared memory and copy the images into it.

Reply:
Shared memory needs to be allocated before this since the image data needs to be copied into the right place here and it requires the pointer from mmap

Reply:
Can't we copy it below just before calling SetSharedMemory?

Reply:
Yes we can